### PR TITLE
Regenerate headless schema on app redeploy or site change

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

File: src/main/java/com/enonic/lib/guillotine/Synchronizer.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.enonic.lib.guillotine;
+
+import com.enonic.xp.script.ScriptValue;
+
+public class Synchronizer
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/guillotine/Synchronizer.java:5]: IMO synchronisation is broken. 
Invalidation happens without this class involved.
Either guard all modifications of map by it or don't synchronise at all

**Review by rymsha — CHANGES_REQUESTED** [review]

File: src/main/java/com/enonic/lib/guillotine/Synchronizer.java (hunk 1 above)
Comment (anatol-sialitski) [reply src/main/java/com/enonic/lib/guillotine/Synchronizer.java:5]: Thank you. I've added sync for invalidate

**Review by rymsha — APPROVED** [review]

**Review by rymsha — APPROVED** [review]

File: src/main/resources/lib/guillotine/index.js [hunk 2]
```diff
@@ -6,6 +6,114 @@ const rootQueryLib = require('/lib/guillotine/query/root-query');
...
+}
+
+var schemaMap = {};
+
+eventLib.listener({
+    type: 'application',
+    localOnly: false,
+    callback: function (event) {
+        if ('STOPPED' === event.data.eventType || 'STARTED' === event.data.eventType) {
+            invalidate();
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:37]: what does it invalidate?

File: src/main/resources/lib/guillotine/index.js [hunk 3]
```diff
@@ -6,6 +6,114 @@ const rootQueryLib = require('/lib/guillotine/query/root-query');
...
+
+function getSiteConfig(siteId, branch) {
+    return contextLib.run({
+        branch: branch
+    }, () => contentLib.getSiteConfig({key: siteId}));
+}
+
+function invalidate(schemaId) {
+    Java.type('com.enonic.lib.guillotine.Synchronizer').sync(__.toScriptValue(function () {
+        if (schemaId) {
```
Comment (rymsha) [inline src/main/resources/lib/guillotine/index.js:99]: getSchama arguments are siteId and branch. Here it should be symmetrical

**Review by alansemenov — APPROVED** [review]

**Review by rymsha — CHANGES_REQUESTED** [review]

**Review by rymsha — APPROVED** [review]

**Pull request merged**